cpp/tests/interop/nanoarrow_utils.hpp

```diff
@@ -135,7 +137,193 @@ void populate_dict_from_col(ArrowArray* arr, cudf::dictionary_column_view dview)
   populate_from_col<KEY_TYPE>(arr->dictionary, dview.keys());
 }
 
+using vector_of_columns = std::vector<std::unique_ptr<cudf::column>>;
+
 std::tuple<std::unique_ptr<cudf::table>, nanoarrow::UniqueSchema, nanoarrow::UniqueArray>
 get_nanoarrow_tables(cudf::size_type length = 10000);
 
 void populate_list_from_col(ArrowArray* arr, cudf::lists_column_view view);
+
+std::unique_ptr<cudf::table> get_cudf_table();
+
+template <typename T>
+struct nanoarrow_storage_type {};
+
+#define DEFINE_NANOARROW_STORAGE(T, NanoType)                    \
+  template <>                                                    \
+  struct nanoarrow_storage_type<T> {                             \
+    static constexpr ArrowType type = NANOARROW_TYPE_##NanoType; \
+  }
+
+DEFINE_NANOARROW_STORAGE(bool, BOOL);
+DEFINE_NANOARROW_STORAGE(int64_t, INT64);
+DEFINE_NANOARROW_STORAGE(uint16_t, UINT16);
+DEFINE_NANOARROW_STORAGE(uint64_t, UINT64);
+DEFINE_NANOARROW_STORAGE(cudf::duration_D, INT32);
+DEFINE_NANOARROW_STORAGE(cudf::duration_s, INT64);
+DEFINE_NANOARROW_STORAGE(cudf::duration_ms, INT64);
+DEFINE_NANOARROW_STORAGE(cudf::duration_us, INT64);
+DEFINE_NANOARROW_STORAGE(cudf::duration_ns, INT64);
+DEFINE_NANOARROW_STORAGE(uint8_t, UINT8);
+DEFINE_NANOARROW_STORAGE(int32_t, INT32);
+
+#undef DEFINE_NANOARROW_STORAGE
+
+template <typename T>
+std::enable_if_t<cudf::is_fixed_width<T>() and !std::is_same_v<T, bool>, nanoarrow::UniqueArray>
+get_nanoarrow_array(std::vector<T> const& data, std::vector<uint8_t> const& mask = {})
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_THROW_NOT_OK(ArrowArrayInitFromType(tmp.get(), nanoarrow_storage_type<T>::type));
+
+  if (!mask.empty()) {
+    ArrowBitmap bitmap;
+    ArrowBitmapInit(&bitmap);
+    NANOARROW_THROW_NOT_OK(ArrowBitmapReserve(&bitmap, mask.size()));
+    ArrowBitmapAppendInt8Unsafe(&bitmap, reinterpret_cast<const int8_t*>(mask.data()), mask.size());
+
+    ArrowArraySetValidityBitmap(tmp.get(), &bitmap);
+    tmp->null_count =
+      data.size() -
+      ArrowBitCountSet(ArrowArrayValidityBitmap(tmp.get())->buffer.data, 0, mask.size());
+  }
+
+  ArrowBuffer buf;
+  ArrowBufferInit(&buf);
+  NANOARROW_THROW_NOT_OK(
+    ArrowBufferAppend(&buf, reinterpret_cast<void const*>(data.data()), sizeof(T) * data.size()));
+  NANOARROW_THROW_NOT_OK(ArrowArraySetBuffer(tmp.get(), 1, &buf));
+
+  tmp->length = data.size();
+
+  return tmp;
```

Comment:
I'm not quite sure how this is working right now. UniqueArrays aren't copyable, are they? So how does returning this local object work here without an explicit `std::move`? Are we relying on copy-elision to happen here? That isn't guaranteed in this scenario where `tmp` is a locally declared variable; C++17 only guarantees it when returning a temporary. Am I missing something, or are we just getting lucky here because of some implementation-specific behavior of the chosen host compiler?

The same question applies to all of the SFINAE overloads of this function.

Reply:
We do rely on copy-elision like this in many places in libcudf. 
Since a unique-ptr object cannot be copied, the compiler has no choice but report an error if copy-elision is not possible.

Reply:
Fair enough, I guess it's OK to rely on the optional behavior behavior in this kind of scenario since there isn't anything that can go wrong other than a failure to compile, which we would see immediately. No risk of UB.